### PR TITLE
Refactor PyVmomiHelper into a base class within module_utils/vmware.py

### LINKED ISSUE #25011: vmware_guest: VM Deployment fails when Datacenter is nested under a folder
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
vmware_guest

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```
also tested against current devel and with #24768

##### CONFIGURATION
None

##### OS / ENVIRONMENT
Ansible on Centos 7
Managing VCenter 6.5

##### SUMMARY
In an enterprise environment, it is common for a VCenter to contain multiple Datacenters, and for these to be organized under various folders instead of the root branch.  When attempting to deploy a VM in this scenario, it will fail.

> Folder
>> DataCenter
>>> Cluster
>>>> Folder
>>>>> VMName

The particular issue is that this piece of code uses the Datacenter name as the start of the path to search for a folder.  
`f_obj = self.content.searchIndex.FindByInventoryPath('/%(datacenter)s%(folder)s' % self.params)`
The Datacenter itself is not the root in this case, and thus we need to append the folder name before it in the search path.

The first way to attempt to work around this was to add the path to the Datacenter name itself, that will then cause this piece of code to fail.
`datacenter = find_obj(self.content, [vim.Datacenter], self.params['datacenter'])`
as it is searching for the exact Datacenter name (and not the full path)

I can then work around it by changing it to this
`datacenter = find_obj(self.content, [vim.Datacenter], self.params['datacenter'].split('/')[-1])`

While this works in my lab (but not at the client site) that does not seem like the proper way to fix this though.  We should ideally have another parameter for the base path of the Datacenter, but that now also adds more complexity.  Another option is to pull the Datacenter first, and then pull its full folder Path, and then use that for searching.

Once a proper decision is made on how we should handle this, then I can do a PR for it.

##### STEPS TO REPRODUCE

```yaml
---
- name: Check VM Info
  hosts: all
  connection: local
  gather_facts: no
  tasks:
  - name: Power On
    vmware_guest:
      hostname: "{{ vcenter_hostname }}"
      username: "{{ vcenter_user }}"
      password: "{{ vcenter_pass }}"
      name: "TestVM"
      state: poweredon
      datacenter: "MyDatacenter"
      cluster: "MyCluster"
      validate_certs: no
      folder: /
      template: CentosBase
```

##### EXPECTED RESULTS
VM is Created in the Nested Datacenter

##### ACTUAL RESULTS

```
fatal: [vcenter]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "annotation": null,
            "cluster": "MyCluster",
            "customization": {},
            "customvalues": [],
            "datacenter": "MyDatacenter",
            "disk": [],
            "esxi_hostname": null,
            "folder": "/vm",
            "force": false,
            "guest_id": null,
            "hardware": {},
            "hostname": "vcenter",
            "is_template": false,
            "name": "TestVM",
            "name_match": "first",
            "networks": [],
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "resource_pool": null,
            "snapshot_src": null,
            "state": "poweredon",
            "template": "CentosBase",
            "template_src": "CentosBase",
            "username": "administrator",
            "uuid": null,
            "validate_certs": false,
            "wait_for_ip_address": false
        }
    },
    "msg": "No folder matched the path: /vm"
}

```

### PR DESCRIPTION
**Pull request by jctanner**
##### SUMMARY
Refactor PyVmomiHelper into a base class within module_utils/vmware.py

* clean up pep8 formatting
* fix various bugs in vm search
* do not munge the folder path given by the user
* fallback to a broad search and narrow down by paths

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
vmware_guest.py vmware.py

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

Fixes #25011

### CODE REVIEW

File: lib/ansible/module_utils/vmware.py [hunk 1]
```diff
@@ -127,7 +242,10 @@ def find_datastore_by_name(content, datastore_name):
 
 def find_dvs_by_name(content, switch_name):
 
-    vmware_distributed_switches = get_all_objs(content, [vim.dvs.VmwareDistributedVirtualSwitch])
+    vmware_distributed_switches = get_all_objs(
+        content, [vim.dvs.VmwareDistributedVirtualSwitch]
+    )
```
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:247]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.

File: lib/ansible/module_utils/vmware.py [hunk 2]
```diff
@@ -275,7 +433,8 @@ def list_snapshots_recursively(snapshots):
     snapshot_data = []
     for snapshot in snapshots:
         snapshot_data.append(deserialize_snapshot_obj(snapshot))
-        snapshot_data = snapshot_data + list_snapshots_recursively(snapshot.childSnapshotList)
+        snapshot_data = snapshot_data + \
+            list_snapshots_recursively(snapshot.childSnapshotList)
```
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:445]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.

File: lib/ansible/module_utils/vmware.py [hunk 3]
```diff
@@ -293,9 +453,13 @@ def list_snapshots(vm):
     if vm.snapshot is None:
         return result
 
-    result['snapshots'] = list_snapshots_recursively(vm.snapshot.rootSnapshotList)
+    result['snapshots'] = list_snapshots_recursively(
+        vm.snapshot.rootSnapshotList
+    )
```
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:466]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.

File: lib/ansible/module_utils/vmware.py [hunk 4]
```diff
@@ -304,10 +468,18 @@ def list_snapshots(vm):
...
+        ),
+        username=dict(
+            type='str', aliases=['user', 'admin'], required=True
+        ),
+        password=dict(
+            type='str', aliases=['pass', 'pwd'], required=True, no_log=True
+        ),
+        validate_certs=dict(
+            type='bool', required=False, default=True
+        ),
```
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:490]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.

File: lib/ansible/module_utils/vmware.py [hunk 5]
```diff
@@ -319,24 +491,36 @@ def connect_to_api(module, disconnect_atexit=True):
     validate_certs = module.params['validate_certs']
 
     if validate_certs and not hasattr(ssl, 'SSLContext'):
-        module.fail_json(msg='pyVim does not support changing verification mode with python < 2.7.9. Either update '
-                             'python or or use validate_certs=false')
+        msg = 'pyVim does not support changing verification mode with'
+        msg += ' python < 2.7.9. Either update python or or use'
+        msg += ' validate_certs=false'
+        module.fail_json(msg=msg)
```
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:505]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.
I really prefer the former over the latter.

File: lib/ansible/module_utils/vmware.py [hunk 6]
```diff
@@ -456,15 +669,19 @@ def push_file_to_guest(content, vm, username, password, src, dest, overwrite=Tru
...
     result = {'failed': False}
 
     tools_status = vm.guest.toolsStatus
     if (tools_status == 'toolsNotInstalled' or
-                tools_status == 'toolsNotRunning'):
+            tools_status == 'toolsNotRunning'):
         result['failed'] = True
-        result['msg'] = "VMwareTools is not installed or is not running in the guest"
+        result['msg'] = "VMwareTools is not installed"
+        result['msg'] = " or is not running in the guest"
```
Comment (dagwieers) [inline lib/ansible/module_utils/vmware.py:692]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.

File: lib/ansible/modules/cloud/vmware/vmware_guest_facts.py [hunk 7]
```diff
@@ -209,7 +135,9 @@ def main():
             e = get_exception()
             module.fail_json(msg="Fact gather failed with exception %s" % e)
     else:
-        module.fail_json(msg="Unable to gather facts for non-existing VM %(name)s" % module.params)
+        msg = "Unable to gather facts for non-existing VM"
+        msg += " %s in folder %s" % (module.params['name'], searchpath)
+        module.fail_json(msg=msg)
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_facts.py:140]: This should not be necessary for PEP8. We have a line-length of 160, not the default 80.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/vmware.py [hunk 8]
```diff
@@ -32,6 +35,117 @@
...
+                recurse=True,
+                allow_duplicates=True
+            )
+            if vms:
+                # return if only one name because we don't have to worry
+                # about duplicates
+                if len(vms) == 1:
+                    return vms[0]
+                else:
+                    # need to find the best path for the duplicate VMs
```
Comment (dericcrago) [inline lib/ansible/module_utils/vmware.py:115]: I appreciate the friendliness of this, but as a user if I've got more than one VM named ``foo`` and I can't give you the right path to it, I think that should fail as a reality check for me.

Comment (dagwieers) [reply lib/ansible/module_utils/vmware.py:115]: I would definitely return a warning back to the user if there are duplicate entries.
The same problem exists when e.g. having a vSwitch and vDS with the samen name.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dav1x** [comment]
shipit

File: lib/ansible/module_utils/vmware.py [hunk 9]
```diff
@@ -32,6 +35,117 @@
...
+
+        # prefer a UUID
+        if uuid is not None:
+            vm = find_vm_by_id(self.content, uuid, vm_id_type="uuid", instanceUuid=False)
+            return vm
+
+        # use a searchpath + FindByInventoryPath to search by whatever
+        # folder the user thinks it is in.
+        elif searchpath:
+
```
Comment (dericcrago) [inline lib/ansible/module_utils/vmware.py:82]: Do you need to get all of the objects and inspect them or could you go straight to the VM?

```python
elif searchpath and name:
    vm_path = searchpath + '/' + name
    vm = self.content.searchIndex.FindByInventoryPath(vm_path)
```

**Comment by jctanner** [comment]
This needs to be totally redone.

**Pull request closed without merging**